awx/main/tasks/host_indirect.py

```diff
@@ -102,11 +109,55 @@ def fetch_job_event_query(job) -> dict[str, str]:
     return net_job_data
 
 
-@task(queue=get_task_queuename)
-def save_indirect_host_entries(job_id):
-    job = Job.objects.get(id=job_id)
+def save_indirect_host_entries_of_job(job: Job) -> None:
+    "Once we have a job and we know that we want to do indirect host processing, this is called"
     job_event_queries = fetch_job_event_query(job)
     records = build_indirect_host_data(job, job_event_queries)
     IndirectManagedNodeAudit.objects.bulk_create(records)
     job.event_queries_processed = True
+
+
+@task(queue=get_task_queuename)
+def save_indirect_host_entries(job_id: int, wait_for_events: bool = True) -> None:
+    try:
+        job = Job.objects.get(id=job_id)
+    except Job.DoesNotExist:
+        logger.debug(f'Job {job_id} seems to be deleted, bailing from save_indirect_host_entries')
+        return
+
+    if wait_for_events:
+        # Gate running this task on the job having all events processed, not just EOF or playbook_on_stats
+        current_events = 0
+        for _ in range(10):
+            current_events = job.job_events.count()
+            if current_events >= job.emitted_events:
+                break
+            logger.debug(f'Waiting for job_id={job_id} to finish processing events, currently {current_events} < {job.emitted_events}')
+            time.sleep(0.2)
+        else:
+            logger.warning(f'Event count {current_events} < {job.emitted_events} for job_id={job_id}, delaying processing of indirect host tracking')
+            return
+
+    try:
+        save_indirect_host_entries_of_job(job)
+    except Exception:
+        logger.traceback(f'Error processing indirect host data for job_id={job_id}')
+
+    # Mark job as processed, even if the processing failed
     job.save(update_fields=['event_queries_processed'])
+
+
+@task(queue=get_task_queuename)
+def save_indirect_host_entries_fallback() -> None:
+    if not flag_enabled("FEATURE_INDIRECT_NODE_COUNTING_ENABLED"):
+        return
+
+    job_ct = 0
+    right_now_time = now()
+    window_end = right_now_time - timedelta(seconds=settings.INDIRECT_HOST_QUERY_FALLBACK_MINUTES)
+    window_start = right_now_time - timedelta(days=settings.INDIRECT_HOST_QUERY_FALLBACK_GIVEUP_DAYS)
+    for job in Job.objects.filter(event_queries_processed=False, finished__lte=window_end, finished__gte=window_start).iterator():
```

Comment:
When this runs the first time, it could lead to a large number of processing jobs (all jobs completed in the last three days). Are we ok with that? Most of them will probably end quickly because there are no queries.

Reply:
Well that seems like a good argument for having a limit to the start of the window which I did here. Otherwise that was more of a gut feeling.

But I still sense something to be worried about. This will get _scheduled_ for a lot of jobs that don't need the processing done. The periodic task will run and schedule a job for each unprocessed job. Each should be a no-op with a single write, but that could still be a lot of jobs. Worse, it could temporarily overwhelm the dispatcher, because it can't delay tasks (ideal might be to scramble them with random delays).

I have a thought for this, and I'll work on putting something up. With a different angle, it'll get more intuitive.